wremnants/include/muon_calibration.h

```diff
@@ -355,21 +342,57 @@ class calibration_uncertainty_helper {
             std::cout << "covinvalt\n" << covinvalt << std::endl;
           }
         }
-        
+
         // protect against outliers
+        // if (weight > 0.9998 && weight < 1.0002) {cout << "smearing weight is " << weight << "covd is " << covd << std::endl;}
         res(ivar, idownup) = std::min(weight, maxWeight_);
-
       }
     }
-
+    
     return res;
   }
-
   std::shared_ptr<const HIST> hist_;
   double minGenPt_;
   double maxWeight_;
 
 };
 
+    float smearGenPt(RVec<float> cov, int charge, float pt, float theta) {
+        float sigma2_qop = cov[0];
+        double sigma2_pt = pow((pow(pt, 2) / (charge * sin(theta))), 2) * sigma2_qop;
+        return gRandom -> Gaus(pt, sqrt(sigma2_pt));
+    }
+    
+    double smearGenQop(RVec<float> cov, double qop) {
+        double sigma2_qop = cov[0];
+        return gRandom -> Gaus(qop, sqrt(sigma2_qop));
+    }
 
+    int getGoodGenMuons0IdxInReco(RVec<bool> goodMuons, RVec<bool> goodMuonsByGenTruth) {
+        int first_goodGenMuon_idx_in_goodMuons = 0;
+        for (int i = 0; i < goodMuonsByGenTruth.size(); i++) {
+            if (goodMuonsByGenTruth[i]) {break;}
+            else {first_goodGenMuon_idx_in_goodMuons += 1;}
+        }
+        int goodGenMuons0_idx_in_recos = 0;
+        int num_goodMuons_found = 0;
+        for (int i = 0; i < goodMuons.size(); i++) {
+            if (goodMuons[i]) {
+                num_goodMuons_found += 1;
+                if (num_goodMuons_found == first_goodGenMuon_idx_in_goodMuons + 1) {
+                    break;
+                }
+            }
+            goodGenMuons0_idx_in_recos += 1;
+        }
+        //if (goodMuons.size() > 1) {
```

Comment:
You can add a debug flag if you want, otherwise I'd get rid of commented out code